pennylane/_qubit_device.py

```diff
@@ -112,83 +112,246 @@ def execute(self, queue, observables, parameters=None):
+
+    @staticmethod
+    def states_to_binary(samples, number_of_states):
+        """Convert basis states from base 10 to binary representation.
+
+        This is an auxiliary method to the generate_samples method.
+
+        Args:
+            samples (List[int]): samples of basis states in base 10 representation
+            number_of_states (int): the number of basis states to sample from
+
+        Returns:
+            List[int]: basis states in binary representation
+        """
+        powers_of_two = (1 << np.arange(number_of_states))
+        states_sampled_base_ten = samples[:, None] & powers_of_two
+        return (states_sampled_base_ten > 0).astype(int)
 
     def expval(self, observable):
         wires = observable.wires
-        self.rotate_basis(observable)
+
+        # TODO: remove the following part as we can assume that
+        # rotate_basis has already been called. Doing so will, however break
+        # unit tests directly calling on expval()
+        if self._rotated_prob is None:
+            self.rotate_basis([observable])
 
         if self.analytic:
             # exact expectation value
             eigvals = observable.eigvals
-            prob = np.abs(self._state ** 2)
-            prob = self.marginal_prob(prob, wires=wires)
+            prob = self.probability(wires=wires)
             return (eigvals @ prob).real
 
         # estimate the ev
         return np.mean(self.sample(observable))
 
     def var(self, observable):
         wires = observable.wires
-        self.rotate_basis(observable)
+
+        # TODO: remove the following part as we can assume that
+        # rotate_basis has already been called. Doing so will, however break
+        # unit tests directly calling on var()
+        if self._rotated_prob is None:
+            self.rotate_basis([observable])
 
         if self.analytic:
             # exact variance value
             eigvals = observable.eigvals
-            prob = np.abs(self._state ** 2)
-            prob = self.marginal_prob(prob, wires=wires)
+            prob = self.probability(wires=wires)
             return (eigvals ** 2) @ prob - (eigvals @ prob).real ** 2
 
+        # estimate the variance
         return np.var(self.sample(observable))
 
+    def sample(self, observable):
+        wires = observable.wires
+        name = observable.name
+
+        # TODO: remove the following part as we can assume that
+        # rotate_basis has already been called. Doing so will, however break
+        # unit tests directly calling on sample()
+        if self._rotated_prob is None:
+            observable.return_type = Sample
+            self.rotate_basis([observable])
+
+            if self._memory or (not self.analytic):
+                self.generate_samples()
 
-    def probability(self, wires=None, values_only=False):
+        if isinstance(name, str) and name in {"PauliX", "PauliY", "PauliZ", "Hadamard"}:
```

Comment:
Is there such a dramatic increase in performance by having the Pauli expvals in a separate (more efficient?) method?

Reply:
Okay, when looking at the code, I guess there is!